File: CaseStudies/FileInputTests/Unbundled/Design3/main.swift
// Unbundled

import Foundation

struct tempParameters {
    let name: String
    let height: Int
    let weight: Double
}

func get_input(_ filename: String) throws -> tempParameters {
    
    let lines = try String(contentsOfFile: filename, encoding: String.Encoding.utf8)
        .components(separatedBy: "\n")
    
    let name: String = lines[1]
    let height: Int = Int(lines[3]) ?? 0
    let weight: Double = Double(lines[5]) ?? 0

    return tempParameters(name:name, height:height, weight:weight)
}

func input_constraints(_ name: String,_ height: Int,_ weight: Double) throws -> Void {
    
    if name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
        throw ParamError("Name must not be blank!")
    }
    else if height <= 0 {
        throw ParamError("Height must be positive!")
    }
    else if weight <= 0 {
        throw ParamError("Weight must be positive!")
    }
}

// Simple error struct
struct ParamError: LocalizedError {
    let description: String

    init(_ description: String) {
        self.description = description
    }

    var errorDescription: String? {
        description
    }
}

let p: tempParameters
p = try get_input("FileInputTests/input.txt")

// This is a bit hacky again :(
let name: String, height: Int, weight: Double
name = p.name
height = p.height
weight = p.weight

try input_constraints(name, height, weight)
print(p)

File: CaseStudies/FileInputTests/Unbundled/Design3/main.jl
# Unbundled
#=
    I like the key-value pairs, but I miss the type-safety of tuples.
    Also, there's still no help from the compiler for knowing that a key exists.
=#

struct tempParameters
    name::String
    height::Integer
    weight::AbstractFloat
end

function get_input(filename::AbstractString)::tempParameters # Loss of type safety unfortunately
    
    file = open(filename, "r")
    lines = readlines(file)
    close(file)

    name = lines[2]
    height = parse(Int32, lines[4])
    weight = parse(Float64, lines[6])

    return tempParameters(name, height, weight)
end

function input_constraints(name::AbstractString, height::Integer, weight::AbstractFloat)::Nothing

    if isempty(strip(name))
        throw("Name must not be blank!")
    elseif height <= 0
        throw("Height must be positive!")
    elseif weight <= 0
        throw("Weight must be positive!")
    end
end

p = get_input("FileInputTests/input.txt")

# This is a bit clearer of where the values are coming from
(name, height, weight) = (p.name, p.height, p.weight)
input_constraints(name, height, weight)
println(p)